# Biosolids data debugging 

Explore biosolids dataset from EPA 

Author: Sahar H. El Abbadi
March 2024 

In [1]:
# set up 
import pandas as pd 
import pathlib 
from tqdm import tqdm 
import pickle
from utilities import check_all_sic_code
tqdm.pandas()

# Load biosolids data 
# all_biosolids = pd.read_csv(pathlib.PurePath('04_results', 'biosolids_with_ww_permits.csv')) # 
# biosolids_not_sewer = pd.read_csv(pathlib.PurePath('04_results', 'bioslids_not_sewer.csv')) # This dataset was created in 06_check_biosolids_list.py

# relevant column is "NPDES ID" 

In [6]:
# Make a pickle file for biosolids data with a column containing SIC code for each file 

all_biosolids = pd.read_csv(pathlib.PurePath('01_raw_data', 'Data_Download_1699657092121.csv'))

# test on top row 
# all_biosolids = all_biosolids.head(2).copy()
all_biosolids['sic_permit'] = all_biosolids['NPDES ID'].progress_apply(check_all_sic_code)
all_biosolids.to_pickle(pathlib.PurePath('05_pickle_files', 'biosolids_data_sic_codes.pkl'))

100%|██████████| 4182/4182 [4:34:18<00:00,  3.94s/it]     


In [7]:
all_biosolids.to_csv((pathlib.PurePath('04_results', 'biosolids_data_sic_codes.csv')))

In [4]:
tqdm.pandas()
from utilities import check_for_ww_permits

# Note: copied from 06_check_biosolids_list.py for the sake of having everything easily accessible in one notebook 

# Load data
biosolids_data = pd.read_csv(pathlib.PurePath('01_raw_data', 'Data_Download_1699657092121.csv'))
biosolids_data['check_sewer_permits'] = biosolids_data['NPDES ID'].progress_apply(check_for_ww_permits)

# Save dataset that has all entries in biosolids dataset, with new column indicating if it has a sewer permit 
biosolids_data.to_csv(pathlib.PurePath('04_results', 'biosolids_with_ww_permits.csv'))

# Filter by facilities that do not have 
biosolids_not_sewer = biosolids_data[biosolids_data['check_sewer_permits'] == 'other_system'].copy()
biosolids_not_sewer.to_csv(pathlib.PurePath('04_results', 'bioslids_not_sewer.csv'))
print('Number of NPDES permits in the bioslids dataset without associated sewer facilities:')
print(f'{len(biosolids_not_sewer)}')

100%|██████████| 4182/4182 [14:26<00:00,  4.83it/s]

Number of NPDES permits in the bioslids dataset without associated sewer facilities:
1110


# Biosolids dataset 

1. Load the biosolids data that does not have a sewer-related permit 
2. Match NPDES permits with all SIC codes 

In [4]:
# Check SIC code for permits, using the database of all SIC codes 

from utilities import check_all_sic_code
# Check for DW permits 
tqdm.pandas()

# Load biosolids dataset filtered to only contain 
biosolids_not_sewer['sic_permits'] = biosolids_not_sewer['NPDES ID'].progress_apply(check_all_sic_code)
display(biosolids_not_sewer)
biosolids_not_sewer.to_csv(pathlib.PurePath('04_results', 'biosolids_not_sewer_all_sic_codes.csv'))

100%|██████████| 1110/1110 [1:00:46<00:00,  3.28s/it]


Unnamed: 0                 Facility Name   NPDES ID  \
0              2                  HANFORD WWTF  CAL001076   
1             14    AUSTIN COUNTY WSC  PLANT 3  TX0125709   
2             15        LAKE PFLUGERVILLE WWTF  TX0132721   
3             16  PURTIS CREEK STATE PARK WWTP  TX0082856   
4             25            CHISOS BASIN  WWTP  TX0094684   
...          ...                           ...        ...   
1105        4157          PINOLE/HERCULES WPCP  CAL037796   
1106        4164            SOUTH WINDSOR WPCF  CTL100510   
1107        4168                PATTERSON WQCF  CAL078735   
1108        4171     NOONDAY WATER WELL 2 WWTF  TX0118362   
1109        4176                RUSHVILLE WWTP  INL021270   

                        City State  EPA Region  Reporting Year  \
0                    HANFORD    CA           9            2022   
1                  BELLVILLE    TX           6            2022   
2               PFLUGERVILLE    TX           6            2022   
3                     AUSTIN    TX           6            2022   
4     BIG BEND NATIONAL PARK    TX           6            2022   
...                      ...   ...         ...             ...   
1105                  PINOLE    CA           9            2022   
1106           SOUTH WINDSOR    CT           1            2022   
1107               PATTERSON    CA           9            2022   
1108                   TYLER    TX           6            2022   
1109               RUSHVILLE    IN           5            2022   

                    Certified Date  \
0     2023-02-10T12:15:23.575-0500   
1     2022-09-07T10:35:30.119-0500   
2     2022-10-07T13:37:35.841-0500   
3     2022-08-01T11:19:59.102-0500   
4     2022-09-14T07:22:25.560-0500   
...                            ...   
1105  2023-06-01T13:28:11.097-0500   
1106  2023-07-26T11:19:07.287-0500   
1107  2023-09-27T18:14:56.033-0500   
1108  2023-09-19T14:18:14.733-0500   
1109  2023-10-16T10:37:43.865-0500   

                                Reporting Obligation(s)  \
0     A Class I Sludge Management Facility as define...   
1                                                 Other   
2     A Class I Sludge Management Facility as define...   
3                                                 Other   
4                                                 Other   
...                                                 ...   
1105  A POTW with a design flow rate equal to or gre...   
1106  A POTW with a design flow rate equal to or gre...   
1107  A POTW with a design flow rate equal to or gre...   
1108           A POTW that serves 10,000 people or more   
1109  A POTW with a design flow rate equal to or gre...   

      Amount of Biosolids Generated  Violation Count Violation Type(s)  \
0                          676.9000                0               NaN   
1                            0.0000                0               NaN   
2                            0.0000                0               NaN   
3                            0.0005                0               NaN   
4                            0.3750                0               NaN   
...                             ...              ...               ...   
1105                       473.0000                0               NaN   
1106                       875.0000                0               NaN   
1107                       286.7800                0               NaN   
1108                         0.0000                0               NaN   
1109                        98.0800                0               NaN   

     Management Practice Type(s)  Amount of Biosolids Managed - Land Applied  \
0               Land Application                                      678.00   
1      Other Management Practice                                         NaN   
2      Other Management Practice                                         NaN   
3      Other Management Practice                                         NaN   
4      Other Management P

/var/folders/ys/gw7hlz293d75kbjkjr6jypth0000gp/T/ipykernel_26618/459731801.py:10: FutureWarning: Starting with pandas version 3.0 all arguments of to_csv except for the argument 'path_or_buf' will be keyword-only.
  biosolids_not_sewer.to_csv('04_results', 'biosolids_not_sewer_all_sic_codes.csv')


IsADirectoryError: [Errno 21] Is a directory: '04_results'

In [7]:
# Summarize findings 
def summarize_sic_codes(permit_list, col_name): 
    unique_entries_count = permit_list[col_name].apply(tuple).nunique()
    
    # Flatten lists and convert them to tuples for consistency
    flat_column = permit_list[col_name].apply(lambda x: tuple(x) if isinstance(x, list) else x)
    
    # Count unique entries and print them along with their counts
    unique_counts = flat_column.value_counts()
    print("Unique entries with counts:")
    print(unique_counts)
    return unique_counts

summarize_sic_codes(biosolids_not_sewer, 'sic_permits')

display(biosolids_not_sewer)

Unique entries with counts:
sic_permits
NO_SIC_MATCH                                  810
(6515,)                                        47
(4941,)                                        35
(8211, 8211)                                   34
(4941, 4941)                                   14
                                             ... 
(7992, 7999)                                    1
(3511, 3511, 7699, 3511, 7699, 3511, 7699)      1
(6531,)                                         1
(8063, 8063)                                    1
(9511, 9511, 1629, 1629, 1629)                  1
Name: count, Length: 111, dtype: int64


Unnamed: 0                 Facility Name   NPDES ID  \
0              2                  HANFORD WWTF  CAL001076   
1             14    AUSTIN COUNTY WSC  PLANT 3  TX0125709   
2             15        LAKE PFLUGERVILLE WWTF  TX0132721   
3             16  PURTIS CREEK STATE PARK WWTP  TX0082856   
4             25            CHISOS BASIN  WWTP  TX0094684   
...          ...                           ...        ...   
1105        4157          PINOLE/HERCULES WPCP  CAL037796   
1106        4164            SOUTH WINDSOR WPCF  CTL100510   
1107        4168                PATTERSON WQCF  CAL078735   
1108        4171     NOONDAY WATER WELL 2 WWTF  TX0118362   
1109        4176                RUSHVILLE WWTP  INL021270   

                        City State  EPA Region  Reporting Year  \
0                    HANFORD    CA           9            2022   
1                  BELLVILLE    TX           6            2022   
2               PFLUGERVILLE    TX           6            2022   
3                     AUSTIN    TX           6            2022   
4     BIG BEND NATIONAL PARK    TX           6            2022   
...                      ...   ...         ...             ...   
1105                  PINOLE    CA           9            2022   
1106           SOUTH WINDSOR    CT           1            2022   
1107               PATTERSON    CA           9            2022   
1108                   TYLER    TX           6            2022   
1109               RUSHVILLE    IN           5            2022   

                    Certified Date  \
0     2023-02-10T12:15:23.575-0500   
1     2022-09-07T10:35:30.119-0500   
2     2022-10-07T13:37:35.841-0500   
3     2022-08-01T11:19:59.102-0500   
4     2022-09-14T07:22:25.560-0500   
...                            ...   
1105  2023-06-01T13:28:11.097-0500   
1106  2023-07-26T11:19:07.287-0500   
1107  2023-09-27T18:14:56.033-0500   
1108  2023-09-19T14:18:14.733-0500   
1109  2023-10-16T10:37:43.865-0500   

                                Reporting Obligation(s)  \
0     A Class I Sludge Management Facility as define...   
1                                                 Other   
2     A Class I Sludge Management Facility as define...   
3                                                 Other   
4                                                 Other   
...                                                 ...   
1105  A POTW with a design flow rate equal to or gre...   
1106  A POTW with a design flow rate equal to or gre...   
1107  A POTW with a design flow rate equal to or gre...   
1108           A POTW that serves 10,000 people or more   
1109  A POTW with a design flow rate equal to or gre...   

      Amount of Biosolids Generated  Violation Count Violation Type(s)  \
0                          676.9000                0               NaN   
1                            0.0000                0               NaN   
2                            0.0000                0               NaN   
3                            0.0005                0               NaN   
4                            0.3750                0               NaN   
...                             ...              ...               ...   
1105                       473.0000                0               NaN   
1106                       875.0000                0               NaN   
1107                       286.7800                0               NaN   
1108                         0.0000                0               NaN   
1109                        98.0800                0               NaN   

     Management Practice Type(s)  Amount of Biosolids Managed - Land Applied  \
0               Land Application                                      678.00   
1      Other Management Practice                                         NaN   
2      Other Management Practice                                         NaN   
3      Other Management Practice                                         NaN   
4      Other Management P

# SIC Code Lookup 

Look up the most common SIC matches
- 6515 = Residential Mobile home site 
- 4941 = Water supply 
- 8211 = Elementary and secondary schools 
- 7992 = Golf course 
- 7033 = Recreational vehicle parks / campsites 
- 7032 = Sports and recreational camping 
- 9223 = Correctional facilities 
- 7033 = Recreational vehicle parks / campsites 
- 9223 = Correctional intstutions
- 7999 = Amusement / Recreation 


In [8]:
# Remove facilities except those without an SIC match 

biosolids_to_remove = biosolids_not_sewer[biosolids_not_sewer['sic_permits'] != ['NO_SIC_MATCH']]
display(biosolids_to_remove)
biosolids_to_remove.describe()
biosolids_to_remove.to_csv(pathlib.PurePath('04_results', 'biosolids_no_sewer_code.csv'))

Unnamed: 0                   Facility Name   NPDES ID  \
1             14      AUSTIN COUNTY WSC  PLANT 3  TX0125709   
2             15          LAKE PFLUGERVILLE WWTF  TX0132721   
3             16    PURTIS CREEK STATE PARK WWTP  TX0082856   
4             25              CHISOS BASIN  WWTP  TX0094684   
5             35     LAUGHLIN AFB WWTP BLDG 1004  TX0022608   
...          ...                             ...        ...   
1019        3966                  CITY OF MONETT  MOL021440   
1095        4133  MERIDIAN METROPOLITAN DISTRICT  COL039110   
1096        4135       BLUE SKY RANCH AND RESORT  UTL025763   
1104        4156                BOLIVAR, CITY OF  MOL022373   
1108        4171       NOONDAY WATER WELL 2 WWTF  TX0118362   

                         City State  EPA Region  Reporting Year  \
1                   BELLVILLE    TX           6            2022   
2                PFLUGERVILLE    TX           6            2022   
3                      AUSTIN    TX           6            2022   
4      BIG BEND NATIONAL PARK    TX           6            2022   
5     LAUGHLIN AIR FORCE BASE    TX           6            2022   
...                       ...   ...         ...             ...   
1019                   MONETT    MO           7            2022   
1095                ENGLEWOOD    CO           8            2022   
1096                  WANSHIP    UT           8            2022   
1104                  BOLIVAR    MO           7            2022   
1108                    TYLER    TX           6            2022   

                    Certified Date  \
1     2022-09-07T10:35:30.119-0500   
2     2022-10-07T13:37:35.841-0500   
3     2022-08-01T11:19:59.102-0500   
4     2022-09-14T07:22:25.560-0500   
5     2022-08-31T07:59:34.879-0500   
...                            ...   
1019  2023-02-27T11:53:52.353-0500   
1095  2023-03-23T10:29:23.333-0500   
1096  2023-04-01T12:56:08.283-0500   
1104  2023-05-22T16:11:11.867-0500   
1108  2023-09-19T14:18:14.733-0500   

                                Reporting Obligation(s)  \
1                                                 Other   
2     A Class I Sludge Management Facility as define...   
3                                                 Other   
4                                                 Other   
5                                                 Other   
...                                                 ...   
1019  A POTW with a design flow rate equal to or gre...   
1095           A POTW that serves 10,000 people or more   
1096  A Class I Sludge Management Facility as define...   
1104  A POTW with a design flow rate equal to or gre...   
1108           A POTW that serves 10,000 people or more   

      Amount of Biosolids Generated  Violation Count Violation Type(s)  \
1                            0.0000                0               NaN   
2                            0.0000                0               NaN   
3                            0.0005                0               NaN   
4                            0.3750                0               NaN   
5                            0.0000                0               NaN   
...                             ...              ...               ...   
1019                       287.0000                0               NaN   
1095                       207.7200                0               NaN   
1096                        54.7200                0               NaN   
1104                       190.4000                0               NaN   
1108                         0.0000                0               NaN   

                      Management Practice Type(s)  \
1                       Other Management Practice   
2                       Other Management Practice   
3                       Other Management Practice   
4                       Other Management Practice   
5                       Other Management Practice   
...                                           ...   
1019              

In [20]:
# look at the facilities generate large amounts of biosolids 
biosolids_to_remove_large = biosolids_to_remove[biosolids_to_remove['Amount of Biosolids Generated'] > 1]

land_application_mask = biosolids_to_remove_large['Management Practice Type(s)'].str.contains('Land Application')
test = biosolids_to_remove_large[land_application_mask].copy()
display(test)

Unnamed: 0                                      Facility Name  \
11            94                                       MEADOWS WWTP   
112          835                  STEPHEN F AUSTIN HISTORICAL PARK    
121          891                                       TURLOCK WQCF   
123          897                              CITY OF RICHMOND SWTP   
145         1017                                 ROLLING RIDGE WWTF   
209         1490                       WOODGATE MOBILE HOME VILLAGE   
210         1491                                  WESTMONT MHP WWTP   
211         1492                                CARBY ROAD MHP WWTP   
212         1496                                RIVER OAKS MHP WWTP   
213         1500                                   SUNDOWN MHP WWTF   
229         1616                         GALVESTON BAY RV PARK WWTF   
230         1630                             WOODTRACE INTERIM WWTP   
252         1763                                RCSD KIMMSWICK WWTP   
305         1906                            FRANKFORT WWTP, CITY OF   
323         1964                                 HERINGTON, CITY OF   
340         2027                                 CITY OF PERRYVILLE   
346         2036                                 CITY OF WASHINGTON   
357         2060                                 CITY OF CALIFORNIA   
381         2138                                     CITY OF JOPLIN   
393         2173                                   CITY OF WATERLOO   
398         2189                               FORT COLLINS CITY OF   
404         2219                                   POUGHKEEPSIE (T)   
408         2227                                WESTMINSTER CITY OF   
423         2258                                WEST LAFAYETTE WWTP   
429         2276                      SANTA FE, CITY OF (AIRPORT RD   
448         2330                      DIXON CORRECTIONAL CENTER STP   
460         2377  COMMANDING OFFICER MARINE CORPS BASE CAMP LEJEUNE   
468         2399                                  RIVERTON, CITY OF   
489         2460                               SCOTT AIR FORCE BASE   
493         2471                                      CITY OF OZARK   
619         2855                                   LOVELAND CITY OF   
635         2903                                        SUFFERN (V)   
639         2911                                 SWIFT BEEF COMPANY   
659         3004                                 BAKERSFIELD WWTP 2   
661         3010                           CHESTERTON WWTP, TOWN OF   
667         3037                                    TOWN OF WALLACE   
677         3071                       PEMBROKE PINES, CITY OF WWTP   
728         3211                                   CITY OF GASTONIA   
747         3277                                 NORTHGLENN CITY OF   
768         3336                            EMWD - SAN JACINTO RWRF   
783         3386                               US MCAS CHERRY POINT   
800         3436              CRANE DIV., NAVAL SURFACE WARFARE CTR   
805         3455                                   BRIGHTON CITY OF   
890         3665                        EMWD - TEMECULA VALLEY RWRF   
891         3667                             CITY OF JEFFERSON CITY   
985         3870                               LYSTEK INTERNATIONAL   
1019        3966                                     CITY OF MONETT   
1095        4133                     MERIDIAN METROPOLITAN DISTRICT   
1096        4135                          BLUE SKY RANCH AND RESORT   
1104        4156                                   BOLIVAR, CITY OF   

       NPDES ID                  City State  EPA Region  Reporting Year  \
11    TX0021270               HOUSTON    TX           6            2022   
112   TX0032549                AUSTIN    TX           6            2022   
121   CAL078948               TURLOCK    CA           9            2022   
123   TX0137600              RICHMOND    TX           6            2022   
145   TX0

# Check all_wwtps facilities removed 

We previously removed facilities from all_wwtps that were from ICIS2022 and did not have any sewer code. Check here what SIC codes those facilities have 

In [10]:
# Check for drinking water permits 
from utilities import check_for_dw_permits

all_wwtps_no_sewer_code = pd.read_csv(pathlib.PurePath('04_results', 'all_wwtps_no_sewer_code_with_SIC.csv'))

dw_mask = all_wwtps_no_sewer_code['SIC_permits'].str.contains('4941')
all_wwtps_no_sewer_code_has_dw_code = all_wwtps_no_sewer_code[dw_mask]
display(all_wwtps_no_sewer_code_has_dw_code)
# all_wwtps_dw_permits = all_wwtps_dw_permits[all_wwtps_dw_permits['SIC_permits'].apply(lambda x: 4941 in x)]

Unnamed: 0.2  Unnamed: 0.1  Unnamed: 0 FACILITY_CODE    SOURCE STATE  \
8            1471          1471        1471       CO00885  CWNS2012    CO   
10           1921          1921        1921       FL01425  CWNS2012    FL   
16           2808          2808        2808       IA02365  CWNS2012    IA   
22           3544          3544        3544       IL00155  CWNS2012    IL   
26           4539          4539        4539       IN03235  CWNS2012    IN   
46           6213          6213        6213       MD01325  CWNS2012    MD   
56           6732          6732        6732       MI03305  CWNS2012    MI   
58           7080          7080        7080       MN02095  ICIS2022    MN   
79           7630          7630        7630       MO05255  CWNS2012    MO   
117          9484          9484        9484       NE04345  CWNS2012    NE   
122          9690          9690        9690       NJ00765  CWNS2012    NJ   
132         10598         10598       10598       OH00035  CWNS2012    OH   
137         10813         10813       10813       OH02195  CWNS2012    OH   
141         10849         10849       10849       OH02555  CWNS2012    OH   
195         11565         11565       11565       OK00935  CWNS2012    OK   
199         12464         12464       12464       PA03135  CWNS2012    PA   
203         12558         12558       12558       PA04075  CWNS2012    PA   
217         13019         13019       13019       PA08685  CWNS2012    PA   
225         13459         13459       13459       TN00175  CWNS2012    TN   
227         13809         13809       13809       TX01035  CWNS2012    TX   
228         13976         13976       13976       TX02685  CWNS2012    TX   
229         14362         14362       14362       TX06350  ICIS2022    TX   
231         14581         14581       14581       TX07670  ICIS2022    TX   
243         14872         14872       14872       UT01095  CWNS2012    UT   
338         16255         16255       16255       WV01525  CWNS2012    WV   

                                         FACILITY            CITY   LATITUDE  \
8                         Rye, Town of Collection             Rye  37.922200   
10                      IRCUD/South Regional WWTF               0  27.587000   
16                        FT MADISON WESTERLY STL    Fort Madison  40.552400   
22                       CANTON OVERFLOW FACILITY               0  40.564200   
26                          Campbell Township RSD               0  39.049400   
46                         EASTERN PRERELEASE STP               0  39.116800   
56                                  ONTONAGON STP               0  46.770800   
58   CRYSTAL LAKE FLOCCULATION TREATMENT FACILITY     ROBBINSDALE  45.021721   
79                                  ELSBERRY WWTP               0  39.162500   
117                                  ULYSSES WWTP               0  40.896800   
122         Montgomery Township - Oxbridge CS/STP      Belle Mead  40.433300   
132                       Arlington WWTP & Sewers               0  40.892000   
137               Loudonville WWTP & Sewer System               0  40.635600   
141                 Aurora Westerly WWTP & Sewers               0  41.307900   
195                                      GORE WWT            Gore  35.561300   
199                              GARRETT BORO STP               0  39.867900   
203              ST FRANCIS COLLEGE - LORETTO STP               0  40.518900   
217               WEST ST CLAIR-PLEASANTVILLE STP               0  40.160600   
225                      Blaine Wastewater System          Blaine  36.159600   
227                             Rancho Viejo WWTP  CAMERON COUNTY  26.042800   
228                              Par Village WWTP               0  31.936200   
229                   CITY OF DALLAS ELM FORK WTP      CARROLLTON  32.972222   
231                HOMER N TANNER JR REGIONAL WTP         AVINGER  32.863611   
243                                JORDANELLE SSD      HEBER CITY  40.634200   
33

In [11]:
# Run all SIC codes on all_wwtps facilities that don't have a sewer code 
all_wwtp_sewer_codes = pd.read_csv(pathlib.PurePath('04_results', 'facilities_with_ww_permits.csv'))

# Only look at facilities that don't have a sewer code - source can be CWNS (since I'm interested in what types of permits these facilities actually have!) 
no_sewer_mask = (all_wwtp_sewer_codes['check_sewer_permits'] == 'other_system' )

# Apply a filter based on source
all_wwtps_no_sewer_code = all_wwtp_sewer_codes[no_sewer_mask]
all_wwtps_no_sewer_code['SIC_permits'] = all_wwtps_no_sewer_code['NPDES_ID'].progress_apply(check_all_sic_code)

all_wwtps_no_sewer_code.to_csv(pathlib.PurePath('04_results', 'all_wwtps_no_sewer_code_with_SIC.csv'))


100%|██████████| 349/349 [19:02<00:00,  3.27s/it]
/var/folders/ys/gw7hlz293d75kbjkjr6jypth0000gp/T/ipykernel_26618/1913298837.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  all_wwtps_no_sewer_code['SIC_permits'] = all_wwtps_no_sewer_code['NPDES_ID'].progress_apply(check_all_sic_code)


# Check facilities removed from all_wwtps

Check what SIC types are associated with the permits for the 21 facilities we decided to remove from all_wwtps. 

all_wwtps_icis_no_sewer_code.csv is the 05_check_icis_facilities.py
- Facilities that are sourced from ICIS 
- The associated NPDES permit is not sewer related 

In [12]:
# Check that the all_wwtps that we're removing from all_wwtps are not just empty SIC codes 

from utilities import check_all_sic_code
from tqdm import tqdm 
from utilities import lookup_sic_code
import pandas as pd 

tqdm.pandas()

all_wwtps_removed = pd.read_csv(pathlib.PurePath('04_results', 'all_wwtps_icis_no_sewer_code.csv'))
print('About to start:')
all_wwtps_removed['SIC_permits'] = all_wwtps_removed['NPDES_ID'].progress_apply(check_all_sic_code)

display(all_wwtps_removed)

all_wwtps_removed.to_csv(pathlib.PurePath('04_results', 'all_wwtps_icis_no_sewer_code_with_SIC.csv'))

About to start:


100%|██████████| 21/21 [01:06<00:00,  3.17s/it]


Unnamed: 0.1  Unnamed: 0 FACILITY_CODE    SOURCE STATE  \
0           2163        2163       CA05060  ICIS2022    CA   
1           2514        2514       FL03815  ICIS2022    FL   
2           5936        5936       KY02865  ICIS2022    KY   
3           5943        5943       LA02955  ICIS2022    LA   
4           5944        5944       LA02960  ICIS2022    LA   
5           7080        7080       MN02095  ICIS2022    MN   
6          11963       11963       OH08865  ICIS2022    OH   
7          14362       14362       TX06350  ICIS2022    TX   
8          14495       14495       TX07220  ICIS2022    TX   
9          14581       14581       TX07670  ICIS2022    TX   
10         14614       14614       TX07940  ICIS2022    TX   
11         14703       14703       TX08570  ICIS2022    TX   
12         14705       14705       TX08580  ICIS2022    TX   
13         14706       14706       TX08585  ICIS2022    TX   
14         14708       14708       TX08595  ICIS2022    TX   
15         14710       14710       TX08605  ICIS2022    TX   
16         14711       14711       TX08610  ICIS2022    TX   
17         14714       14714       TX08625  ICIS2022    TX   
18         14715       14715       TX08630  ICIS2022    TX   
19         14721       14721       TX08660  ICIS2022    TX   
20         14748       14748       TX08870  ICIS2022    TX   

                                        FACILITY              CITY   LATITUDE  \
0                     THUNDER VALLEY CASINO WWTP           LINCOLN  38.839111   
1               MILITARY POINT INDUSTRIAL LAGOON       PANAMA CITY  30.115600   
2                                NEW MARION WWTP            MARION  37.331667   
3                      MINDEN DIESEL POWER PLANT            MINDEN  32.609167   
4                       MINDEN STEAM POWER PLANT            MINDEN  32.603611   
5   CRYSTAL LAKE FLOCCULATION TREATMENT FACILITY       ROBBINSDALE  45.021721   
6                  HEATHER HILL CARE COMMUNITIES           CHARDON  41.536300   
7                    CITY OF DALLAS ELM FORK WTP        CARROLLTON  32.972222   
8                            FISHING HARBOR WWTP       BROWNSVILLE  25.981388   
9                 HOMER N TANNER JR REGIONAL WTP           AVINGER  32.863611   
10      MEXIA STATE SUPPORTED LIVING CENTER WWTP             MEXIA  31.695000   
11                       TDCJ CLEMENS UNIT  WWTP          BRAZORIA  28.992676   
12                          TDCJ ELLIS UNIT WWTP        HUNTSVILLE  30.887500   
13                            TDCJ ESTELLE UNIT         HUNTSVILLE  30.876944   
14                       TDCJ JESTER 3 UNIT WWTP  FORT BEND COUNTY  29.621744   
15                        TDCJ LUTHER  UNIT WWTP          COURTNEY  30.261514   
16                      TDCJ POWLEDGE UNIT  WWTF   ANDERSON COUNTY  31.729444   
17                            TDJC COFFIELD UNIT   ANDERSON COUNTY  31.788611   
18                             TERRELL UNIT WWTP   BRAZORIA COUNTY  29.297777   
19                               TOLEDO BEND DAM        BURKEVILLE  31.173397   
20                 WILBARGER CREEK REGIONAL WWTF      PFLUGERVILLE  30.386388   

     LONGITUDE  FLOW_2012_MGD  2012_TOT_ANNUAL_MM3  CWNS_NUM   NPDES_ID  \
0  -121.307556          0.875             1.209793         0  CA0084697   
1   -85.630533         37.000            51.156977         0  FL0002631   
2   -88.081389          0.000             0.000000         0  KY0113247   
3   -93.290278          0.500             0.691311         0  LA0109886   
4   -93.295000          0.500             0.691311         0  LA0109894   
5   -93.326900          0.720             0.995487         0  MN0069957   
6   -81.225010          0.100             0.138262         0  OH0083984   
7   -96.916388          5.000             6.913105         0  TX0002372   
8   -97.337500          0.250             0.345655         0  TX0100242   
9    94.662222          0.000             0.000000         0  TX0134490   
10  -96.531944          0.450  

In [13]:
display(all_wwtps_removed)
summarize_sic_codes(all_wwtps_removed, 'SIC_permits')


Unnamed: 0.1  Unnamed: 0 FACILITY_CODE    SOURCE STATE  \
0           2163        2163       CA05060  ICIS2022    CA   
1           2514        2514       FL03815  ICIS2022    FL   
2           5936        5936       KY02865  ICIS2022    KY   
3           5943        5943       LA02955  ICIS2022    LA   
4           5944        5944       LA02960  ICIS2022    LA   
5           7080        7080       MN02095  ICIS2022    MN   
6          11963       11963       OH08865  ICIS2022    OH   
7          14362       14362       TX06350  ICIS2022    TX   
8          14495       14495       TX07220  ICIS2022    TX   
9          14581       14581       TX07670  ICIS2022    TX   
10         14614       14614       TX07940  ICIS2022    TX   
11         14703       14703       TX08570  ICIS2022    TX   
12         14705       14705       TX08580  ICIS2022    TX   
13         14706       14706       TX08585  ICIS2022    TX   
14         14708       14708       TX08595  ICIS2022    TX   
15         14710       14710       TX08605  ICIS2022    TX   
16         14711       14711       TX08610  ICIS2022    TX   
17         14714       14714       TX08625  ICIS2022    TX   
18         14715       14715       TX08630  ICIS2022    TX   
19         14721       14721       TX08660  ICIS2022    TX   
20         14748       14748       TX08870  ICIS2022    TX   

                                        FACILITY              CITY   LATITUDE  \
0                     THUNDER VALLEY CASINO WWTP           LINCOLN  38.839111   
1               MILITARY POINT INDUSTRIAL LAGOON       PANAMA CITY  30.115600   
2                                NEW MARION WWTP            MARION  37.331667   
3                      MINDEN DIESEL POWER PLANT            MINDEN  32.609167   
4                       MINDEN STEAM POWER PLANT            MINDEN  32.603611   
5   CRYSTAL LAKE FLOCCULATION TREATMENT FACILITY       ROBBINSDALE  45.021721   
6                  HEATHER HILL CARE COMMUNITIES           CHARDON  41.536300   
7                    CITY OF DALLAS ELM FORK WTP        CARROLLTON  32.972222   
8                            FISHING HARBOR WWTP       BROWNSVILLE  25.981388   
9                 HOMER N TANNER JR REGIONAL WTP           AVINGER  32.863611   
10      MEXIA STATE SUPPORTED LIVING CENTER WWTP             MEXIA  31.695000   
11                       TDCJ CLEMENS UNIT  WWTP          BRAZORIA  28.992676   
12                          TDCJ ELLIS UNIT WWTP        HUNTSVILLE  30.887500   
13                            TDCJ ESTELLE UNIT         HUNTSVILLE  30.876944   
14                       TDCJ JESTER 3 UNIT WWTP  FORT BEND COUNTY  29.621744   
15                        TDCJ LUTHER  UNIT WWTP          COURTNEY  30.261514   
16                      TDCJ POWLEDGE UNIT  WWTF   ANDERSON COUNTY  31.729444   
17                            TDJC COFFIELD UNIT   ANDERSON COUNTY  31.788611   
18                             TERRELL UNIT WWTP   BRAZORIA COUNTY  29.297777   
19                               TOLEDO BEND DAM        BURKEVILLE  31.173397   
20                 WILBARGER CREEK REGIONAL WWTF      PFLUGERVILLE  30.386388   

     LONGITUDE  FLOW_2012_MGD  2012_TOT_ANNUAL_MM3  CWNS_NUM   NPDES_ID  \
0  -121.307556          0.875             1.209793         0  CA0084697   
1   -85.630533         37.000            51.156977         0  FL0002631   
2   -88.081389          0.000             0.000000         0  KY0113247   
3   -93.290278          0.500             0.691311         0  LA0109886   
4   -93.295000          0.500             0.691311         0  LA0109894   
5   -93.326900          0.720             0.995487         0  MN0069957   
6   -81.225010          0.100             0.138262         0  OH0083984   
7   -96.916388          5.000             6.913105         0  TX0002372   
8   -97.337500          0.250             0.345655         0  TX0100242   
9    94.662222          0.000             0.000000         0  TX0134490   
10  -96.531944          0.450  

Unique entries with counts:
SIC_permits
(9223, 9223)                              5
(4941, 4941)                              3
NO_SIC_MATCH                              2
(4911, 4911, 4911, 4911)                  2
(7011, 7011, 7011, 7011, 1522)            1
(2621, 2621, 2611, 2621, 2611, 2611)      1
(8051, 8051)                              1
(4491, 4491)                              1
(8063, 8063)                              1
(9223, 9223, 9223, 9223, 213)             1
(213, 213, 9223, 9223, 213, 252, 9223)    1
(213, 9223, 213, 9223)                    1
(4911, 4911)                              1
Name: count, dtype: int64


SIC_permits
(9223, 9223)                              5
(4941, 4941)                              3
NO_SIC_MATCH                              2
(4911, 4911, 4911, 4911)                  2
(7011, 7011, 7011, 7011, 1522)            1
(2621, 2621, 2611, 2621, 2611, 2611)      1
(8051, 8051)                              1
(4491, 4491)                              1
(8063, 8063)                              1
(9223, 9223, 9223, 9223, 213)             1
(213, 213, 9223, 9223, 213, 252, 9223)    1
(213, 9223, 213, 9223)                    1
(4911, 4911)                              1
Name: count, dtype: int64

In [14]:
# Summarize SIC code count  
summarize_sic_codes(all_wwtps_no_sewer_code, 'SIC_permits')

Unique entries with counts:
SIC_permits
NO_SIC_MATCH                                        208
(4941,)                                              10
(6552, 6552)                                          9
(4941, 4941)                                          9
(9223, 9223)                                          6
                                                   ... 
(6552, 6552, 6552, 6552)                              1
(9512, 4491, 4491)                                    1
(1629, 9512, 1629, 9512)                              1
(9512, 4231, 4231, 9512, 9512, 4231, 7999, 9512)      1
(4941, 4941, 4941)                                    1
Name: count, Length: 88, dtype: int64


SIC_permits
NO_SIC_MATCH                                        208
(4941,)                                              10
(6552, 6552)                                          9
(4941, 4941)                                          9
(9223, 9223)                                          6
                                                   ... 
(6552, 6552, 6552, 6552)                              1
(9512, 4491, 4491)                                    1
(1629, 9512, 1629, 9512)                              1
(9512, 4231, 4231, 9512, 9512, 4231, 7999, 9512)      1
(4941, 4941, 4941)                                    1
Name: count, Length: 88, dtype: int64

In [15]:
# Check EBMUD 

npdes = 'CAL037702'
test = check_all_sic_code(npdes)
print(test)

NO_SIC_MATCH


In [10]:
# Check if all NPDES IDs are unique in this dataset 
import pandas as pd 
import pathlib
all_biosolids = pd.read_pickle(pathlib.PurePath('05_pickle_files', 'biosolids_data_sic_codes.pkl'))

all_biosolids_nonunique = all_biosolids.duplicated(subset=['NPDES ID'], keep=False)
biosolids_non_unique_npdes = all_biosolids[all_biosolids_nonunique]
display(biosolids_non_unique_npdes)

# ALl NPDES IDs are unique! good :) 

Empty DataFrame
Columns: [Facility Name, NPDES ID, City, State, EPA Region, Reporting Year, Certified Date, Reporting Obligation(s), Amount of Biosolids Generated, Violation Count, Violation Type(s), Management Practice Type(s), Amount of Biosolids Managed - Land Applied, Amount of Biosolids Managed - Surface Disposal, Amount of Biosolids Managed - Incinerated, Amount of Biosolids Managed - Other Management Practice, Pathogen Class(es), sic_permit]
Index: []